src/praisonai-agents/praisonaiagents/agent/agent.py

```
@@ -5644,6 +5749,16 @@ def close(self) -> None:
         except Exception as e:
             logger.warning(f"MCP cleanup failed: {e}")
 
+        # Runtime-attached MCP servers cleanup
+        try:
+            if hasattr(self, '_mcp_servers') and self._mcp_servers:
+                for server in self._mcp_servers.values():
+                    if hasattr(server, 'shutdown'):
+                        server.shutdown()
+                self._mcp_servers.clear()
+        except Exception as e:
```

Ruff (0.15.18)

[warning] Do not catch blind exception: `Exception`

(BLE001)